weave/trace_server/file_storage.py

```diff
@@ -0,0 +1,225 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Union, cast
+
+import boto3
+from azure.storage.blob import BlobServiceClient
+from botocore.config import Config
+from google.cloud import storage
+from google.oauth2.credentials import Credentials as GCPCredentials
+from tenacity import (
+    RetryCallState,
+    before_sleep_log,
+    retry,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from weave.trace_server.file_storage_credentials import (
+    AWSCredentials,
+    AzureAccountCredentials,
+    AzureConnectionCredentials,
+    get_aws_credentials,
+    get_azure_credentials,
+    get_gcp_credentials,
+)
+from weave.trace_server.file_storage_uris import (
+    AzureFileStorageURI,
+    FileStorageURI,
+    GCSFileStorageURI,
+    S3FileStorageURI,
+)
+
+# Configure logger
+logger = logging.getLogger(__name__)
+
+# Default timeout values (in seconds)
+DEFAULT_CONNECT_TIMEOUT = 10
+DEFAULT_READ_TIMEOUT = 30
+RETRY_MAX_ATTEMPTS = 3
+RETRY_MIN_WAIT = 1  # seconds
+RETRY_MAX_WAIT = 10  # seconds
+
+# Publicly exposed methods:
+
+
+def store_in_bucket(file_storage_uri: FileStorageURI, data: bytes) -> None:
+    """Store a file in a storage bucket."""
+    try:
+        client = _get_storage_client(file_storage_uri)
+        client.store(file_storage_uri, data)
+    except Exception as e:
+        logger.exception("Failed to store file at %s: %s", file_storage_uri, str(e))
+        raise type(e)(f"Failed to store file at {file_storage_uri}: {str(e)}") from e
+
+
+def read_from_bucket(file_storage_uri: FileStorageURI) -> bytes:
+    """Read a file from a storage bucket."""
+    try:
+        client = _get_storage_client(file_storage_uri)
+        return client.read(file_storage_uri)
+    except Exception as e:
+        logger.exception("Failed to read file from %s: %s", file_storage_uri, str(e))
+        raise type(e)(f"Failed to read file from {file_storage_uri}: {str(e)}") from e
+
+
+### Everything below here is interal
```

Comment:
then the methods should be underscore prefixed and/or you can explicitly specify `__all__`